**How to Query the NYC Open Data (BigQuery Dataset)**

In [1]:
import bq_helper
from bq_helper import BigQueryHelper
# https://www.kaggle.com/sohier/introduction-to-the-bq-helper-package
nyc = bq_helper.BigQueryHelper(active_project="bigquery-public-data",
                                   dataset_name="new_york")

In [2]:
bq_assistant = BigQueryHelper("bigquery-public-data", "new_york")
bq_assistant.list_tables()

['311_service_requests',
 'citibike_stations',
 'citibike_trips',
 'nypd_mv_collisions',
 'tlc_fhv_trips_2015',
 'tlc_fhv_trips_2016',
 'tlc_green_trips_2013',
 'tlc_green_trips_2014',
 'tlc_green_trips_2015',
 'tlc_green_trips_2016',
 'tlc_yellow_trips_2009',
 'tlc_yellow_trips_2010',
 'tlc_yellow_trips_2011',
 'tlc_yellow_trips_2012',
 'tlc_yellow_trips_2013',
 'tlc_yellow_trips_2014',
 'tlc_yellow_trips_2015',
 'tlc_yellow_trips_2016',
 'tree_census_1995',
 'tree_census_2005',
 'tree_census_2015',
 'tree_species']

In [3]:
bq_assistant.head("tlc_yellow_trips_2016", num_rows=10)

vendor_id           pickup_datetime          dropoff_datetime  \
0         1 2016-04-03 10:57:51+00:00 2016-04-03 11:18:59+00:00   
1         1 2016-07-05 00:11:29+00:00 2016-07-05 00:33:31+00:00   
2         2 2016-02-29 08:52:02+00:00 2016-02-29 09:30:42+00:00   
3         2 2016-01-17 16:33:57+00:00 2016-01-17 17:01:37+00:00   
4         1 2016-01-30 00:42:39+00:00 2016-01-30 01:08:33+00:00   
5         2 2016-03-04 13:47:33+00:00 2016-03-04 14:11:52+00:00   
6         2 2016-08-31 17:45:07+00:00 2016-08-31 17:46:16+00:00   
7         2 2016-09-29 02:51:49+00:00 2016-09-29 03:20:25+00:00   
8         2 2016-11-25 19:56:52+00:00 2016-11-25 19:58:12+00:00   
9         2 2016-05-11 12:48:25+00:00 2016-05-11 13:28:19+00:00   

   passenger_count  trip_distance  pickup_longitude  pickup_latitude  \
0                2           7.20        -73.997177        40.716515   
1                1           8.80               NaN              NaN   
2                2           5.84        -73.970963        40.790916   
3                5           6.92        -73.956718        40.780960   
4                1           6.40        -74.005386        40.720814   
5                1           8.32        -73.958992        40.772034   
6                1           0.06               NaN              NaN   
7                2          10.64               NaN              NaN   
8                1           0.31               NaN              NaN   
9                1          12.23        -73.977463        40.759514   

   rate_code store_and_fwd_flag  dropoff_longitude  dropoff_latitude  \
0          1                  N         -73.954987         40.777653   
1          1                  N                NaN               NaN   
2          1                  N         -73.991661         40.731342   
3          1                  N         -74.001740         40.730331   
4          1                  N         -73.929634         40.685833   
5          1                  N         -73.872536         40.774330   
6          1                  N                NaN               NaN   
7          1                  N                NaN               NaN   
8          1                  N                NaN               NaN   
9          1                  N         -73.867607         40.770676   

  payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0            1         23.0    0.0      0.5        4.75          0.00   
1            1         27.0    0.5      0.5        5.65          0.00   
2            1         27.0    0.0      0.5        2.78          0.00   
3            1         25.0    0.0      0.5        5.16          0.00   
4            1         23.0    0.5      0.5        6.05          0.00   
5            1         26.0    0.0      0.5        6.47          5.54   
6            1          3.0    1.0      0.5        0.96          0.00   
7            1         33.0    0.5      0.5        6.86          0.00   
8            1          3.0    1.0      0.5        0.96          0.00   
9            1         40.5    0.0      0.5       11.71          5.54   

   imp_surcharge  total_amount  
0            0.3         28.55  
1            0.3         33.95  
2            0.3         30.58  
3            0.3         30.96  
4            0.3         30.35  
5            0.3         38.81  
6            0.3          5.76  
7            0.3         41.16  
8            0.3          5.76  
9            0.3         58.55

In [4]:
bq_assistant.head("citibike_trips", num_rows=10)

tripduration                 starttime                  stoptime  \
0          3030 2016-06-18 11:54:48+00:00 2016-06-18 12:45:19+00:00   
1          1376 2013-09-04 20:49:30+00:00 2013-09-04 21:12:26+00:00   
2           870 2016-09-28 09:57:29+00:00 2016-09-28 10:11:59+00:00   
3           313 2015-07-09 17:25:53+00:00 2015-07-09 17:31:07+00:00   
4           792 2015-04-02 08:31:11+00:00 2015-04-02 08:44:24+00:00   
5           372 2013-11-06 13:09:51+00:00 2013-11-06 13:16:03+00:00   
6           470 2014-08-05 15:19:30+00:00 2014-08-05 15:27:20+00:00   
7          1661 2016-07-27 09:31:08+00:00 2016-07-27 09:58:49+00:00   
8           612 2015-07-12 19:22:48+00:00 2015-07-12 19:33:01+00:00   
9           373 2014-11-07 10:21:31+00:00 2014-11-07 10:27:44+00:00   

   start_station_id       start_station_name  start_station_latitude  \
0               237          E 11 St & 2 Ave               40.730473   
1              2017          E 43 St & 2 Ave               40.750224   
2              2010     Grand St & Greene St               40.721655   
3               293    Lafayette St & E 8 St               40.730287   
4               536          1 Ave & E 30 St               40.741444   
5               237          E 11 St & 2 Ave               40.730473   
6               410  Suffolk St & Stanton St               40.720664   
7               280          E 10 St & 5 Ave               40.733320   
8               385          E 55 St & 2 Ave               40.757973   
9               223          W 13 St & 7 Ave               40.737815   

   start_station_longitude  end_station_id          end_station_name  \
0               -73.986724             259   South St & Whitehall St   
1               -73.971214             164           E 47 St & 2 Ave   
2               -74.002347             363            West Thames St   
3               -73.990765             250  Lafayette St & Jersey St   
4               -73.975361             293     Lafayette St & E 8 St   
5               -73.986724             433        E 13 St & Avenue A   
6               -73.985180             332                 Cherry St   
7               -73.995101             260      Broad St & Bridge St   
8               -73.966033             499        Broadway & W 60 St   
9               -73.999947             168           W 18 St & 6 Ave   

   end_station_latitude  end_station_longitude  bikeid    usertype  \
0             40.701221             -74.012342   15677  Subscriber   
1             40.753231             -73.970325   19231    Customer   
2             40.708347             -74.017134   21452  Subscriber   
3             40.724561             -73.995653   22165  Subscriber   
4             40.730287             -73.990765   20851  Subscriber   
5             40.729554             -73.980572   19007  Subscriber   
6             40.712199             -73.979481   18036  Subscriber   
7             40.703652             -74.011678   21013  Subscriber   
8             40.769155             -73.981918   22480  Subscriber   
9             40.739713             -73.994564   20840  Subscriber   

   birth_year   gender  
0      1968.0     male  
1         NaN  unknown  
2      1965.0     male  
3      1991.0   female  
4      1988.0     male  
5      1971.0     male  
6      1968.0   female  
7      1941.0     male  
8      1968.0     male  
9      1984.0     male

In [5]:
bq_assistant.table_schema("nypd_mv_collisions")

[SchemaField('borough', 'STRING', 'NULLABLE', '', ()),
 SchemaField('contributing_factor_vehicle_1', 'STRING', 'NULLABLE', '', ()),
 SchemaField('contributing_factor_vehicle_2', 'STRING', 'NULLABLE', '', ()),
 SchemaField('contributing_factor_vehicle_3', 'STRING', 'NULLABLE', '', ()),
 SchemaField('contributing_factor_vehicle_4', 'STRING', 'NULLABLE', '', ()),
 SchemaField('contributing_factor_vehicle_5', 'STRING', 'NULLABLE', '', ()),
 SchemaField('cross_street_name', 'STRING', 'NULLABLE', '', ()),
 SchemaField('timestamp', 'TIMESTAMP', 'NULLABLE', '', ()),
 SchemaField('latitude', 'FLOAT', 'NULLABLE', '', ()),
 SchemaField('longitude', 'FLOAT', 'NULLABLE', '', ()),
 SchemaField('location', 'STRING', 'NULLABLE', '', ()),
 SchemaField('number_of_cyclist_injured', 'INTEGER', 'NULLABLE', '', ()),
 SchemaField('number_of_cyclist_killed', 'INTEGER', 'NULLABLE', '', ()),
 SchemaField('number_of_motorist_injured', 'INTEGER', 'NULLABLE', '', ()),
 SchemaField('number_of_motorist_killed', 'INT

On which New York City streets are you most likely to find a loud party?


In [6]:
query1 = """SELECT
  street_name,
  borough,
  COUNT(*) AS count
FROM
  `bigquery-public-data.new_york.311_service_requests`
WHERE
  descriptor="Loud Music/Party"
  AND complaint_type="Noise - Residential"
  AND street_name != ""
GROUP BY
  street_name,
  borough
ORDER BY
  count DESC
LIMIT
  1000;
        """
response1 = nyc.query_to_pandas_safe(query1)
response1.head(50)

street_name    borough  count
0     GRAND CONCOURSE      BRONX   7668
1       MORRIS AVENUE      BRONX   4119
2            BROADWAY  MANHATTAN   3970
3        OCEAN AVENUE   BROOKLYN   3820
4    AMSTERDAM AVENUE  MANHATTAN   3295
5       WALTON AVENUE      BRONX   3242
6     DAVIDSON AVENUE      BRONX   3117
7   UNIVERSITY AVENUE      BRONX   2996
8      HOLLAND AVENUE      BRONX   2875
9      SHERMAN AVENUE  MANHATTAN   2578
10   VALENTINE AVENUE      BRONX   2492
11     EAST 21 STREET   BROOKLYN   2273
12    SHERIDAN AVENUE      BRONX   2213
13     BEDFORD AVENUE   BROOKLYN   2192
14      BRYANT AVENUE      BRONX   2175
15      TAYLOR AVENUE      BRONX   2159
16      MONUMENT WALK   BROOKLYN   2150
17   LINDEN BOULEVARD   BROOKLYN   2125
18         PARK PLACE   BROOKLYN   2080
19      PUTNAM AVENUE   BROOKLYN   2059
20     CRESTON AVENUE      BRONX   2039
21     DECATUR AVENUE      BRONX   2013
22           8 AVENUE  MANHATTAN   1959
23       GRAND AVENUE      BRONX   1959
24     WEBSTER AVENUE      BRONX   1900
25     ELLWOOD STREET  MANHATTAN   1900
26    RIVERSIDE DRIVE  MANHATTAN   1894
27      DEKALB AVENUE   BROOKLYN   1860
28         LENOX ROAD   BROOKLYN   1813
29      GERARD AVENUE      BRONX   1805
30     STERLING PLACE   BROOKLYN   1795
31   ROCKAWAY PARKWAY   BROOKLYN   1773
32    NEW YORK AVENUE   BROOKLYN   1758
33       UNION STREET   BROOKLYN   1749
34    WEST 141 STREET  MANHATTAN   1732
35     HANCOCK STREET   BROOKLYN   1718
36      HALSEY STREET   BROOKLYN   1708
37           1 AVENUE  MANHATTAN   1704
38         110 STREET     QUEENS   1687
39       BEACH AVENUE      BRONX   1678
40           2 AVENUE  MANHATTAN   1652
41    NOSTRAND AVENUE   BROOKLYN   1646
42      LINCOLN PLACE   BROOKLYN   1629
43     PROSPECT PLACE   BROOKLYN   1628
44     WALLACE AVENUE      BRONX   1622
45    PROSPECT AVENUE      BRONX   1620
46    EASTERN PARKWAY   BROOKLYN   1616
47  WASHINGTON AVENUE      BRONX   1556
48  WOODYCREST AVENUE      BRONX   1553
49    WEST 129 STREET  MANHATTAN   1545

What are the top five reasons why New Yorkers call 311 over the past 4 years?



In [7]:
query2 = """SELECT
  Extract(YEAR from created_date) AS year,
  REPLACE(UPPER(complaint_type), "HEATING", "HEAT/HOT WATER") as complaint, 
  COUNT(*) AS count
FROM
  `bigquery-public-data.new_york.311_service_requests`
GROUP BY complaint, year
ORDER BY COUNT DESC
LIMIT 1000;
        """
response2 = nyc.query_to_pandas_safe(query2, max_gb_scanned=10)
response2.head(15)

year            complaint   count
0   2014       HEAT/HOT WATER  230364
1   2017  NOISE - RESIDENTIAL  230152
2   2016       HEAT/HOT WATER  227959
3   2015       HEAT/HOT WATER  225706
4   2016  NOISE - RESIDENTIAL  221906
5   2010       HEAT/HOT WATER  214218
6   2017       HEAT/HOT WATER  213521
7   2015  NOISE - RESIDENTIAL  208026
8   2013       HEAT/HOT WATER  202896
9   2014  NOISE - RESIDENTIAL  192608
10  2011       HEAT/HOT WATER  190184
11  2012       HEAT/HOT WATER  182974
12  2013  NOISE - RESIDENTIAL  151516
13  2017      ILLEGAL PARKING  146122
14  2017     BLOCKED DRIVEWAY  136097

![](https://cloud.google.com/blog/big-data/2017/01/images/148467900588042/nyc-dataset-1.png)
https://cloud.google.com/blog/big-data/2017/01/images/148467900588042/nyc-dataset-1.png

Can you find the Virginia Pines in New York City?


In [8]:
query3 = """SELECT
  Latitude,
  Longitude
FROM
  `bigquery-public-data.new_york.tree_census_2015`
WHERE
  spc_common="Virginia pine";
        """
response3 = nyc.query_to_pandas_safe(query3, max_gb_scanned=10)
response3.head(15)

Latitude  Longitude
0  40.725547 -73.978741
1  40.724112 -73.985056
2  40.786300 -73.973990
3  40.704530 -73.903470
4  40.692954 -73.794684
5  40.676955 -73.899443
6  40.676982 -73.899192
7  40.667671 -73.987426
8  40.868704 -73.848901
9  40.742036 -73.715848

Where can you find the most of your favorite type of tree?

In [9]:
query3 = """SELECT
  UPPER(spc_common) AS common_name,
  ROUND(latitude, 1) AS lat,
  ROUND(longitude, 1) AS long,
  COUNT(*) AS tree_count
FROM
  `bigquery-public-data.new_york.tree_census_2015`
GROUP BY
  spc_common,
  spc_latin,
  lat,
  long
ORDER BY
  tree_count DESC;
        """
response3 = nyc.query_to_pandas_safe(query3, max_gb_scanned=10)
response3.head(15)

common_name   lat  long  tree_count
0   LONDON PLANETREE  40.7 -73.9       12940
1   LONDON PLANETREE  40.6 -74.0       12528
2   LONDON PLANETREE  40.7 -73.8       11916
3        HONEYLOCUST  40.7 -73.9       11658
4       CALLERY PEAR  40.5 -74.2       11483
5   LONDON PLANETREE  40.6 -73.9       10471
6        HONEYLOCUST  40.8 -73.9       10096
7            PIN OAK  40.7 -73.8        9056
8        HONEYLOCUST  40.7 -74.0        8389
9            PIN OAK  40.7 -73.9        7186
10      NORWAY MAPLE  40.7 -73.8        7075
11  LONDON PLANETREE  40.6 -74.1        6655
12  LONDON PLANETREE  40.8 -73.9        6617
13  LONDON PLANETREE  40.7 -74.0        6527
14       HONEYLOCUST  40.7 -73.8        6249

What are the most popular types of tree in New York City?

In [10]:
query4 = """SELECT
  UPPER(spc_common) as common_name,
  UPPER(spc_latin) as latin_name,
  COUNT(*) AS tree_count
FROM
  `bigquery-public-data.new_york.tree_census_2015`
GROUP BY
  spc_common,
  spc_latin
ORDER BY
  tree_count DESC
LIMIT
  10;
        """
response4 = nyc.query_to_pandas_safe(query4, max_gb_scanned=10)
response4.head(15)

common_name                          latin_name  tree_count
0   LONDON PLANETREE               PLATANUS X ACERIFOLIA       87014
1        HONEYLOCUST  GLEDITSIA TRIACANTHOS VAR. INERMIS       64264
2       CALLERY PEAR                    PYRUS CALLERYANA       58931
3            PIN OAK                   QUERCUS PALUSTRIS       53185
4       NORWAY MAPLE                    ACER PLATANOIDES       34189
5                                                              31619
6  LITTLELEAF LINDEN                       TILIA CORDATA       29742
7             CHERRY                              PRUNUS       29279
8   JAPANESE ZELKOVA                     ZELKOVA SERRATA       29258
9             GINKGO                       GINKGO BILOBA       21024

How many times have animals injured cyclists in New York City?


In [11]:
query5 = """SELECT
  contributing_factor_vehicle_1,
  number_of_cyclist_injured,
  latitude,
  longitude
FROM
  `bigquery-public-data.new_york.nypd_mv_collisions`
WHERE
  STRPOS(contributing_factor_vehicle_1,"Animal") > 0
  AND number_of_cyclist_injured > 0;
        """
response5 = nyc.query_to_pandas_safe(query5, max_gb_scanned=10)
response5.head(15)

contributing_factor_vehicle_1  number_of_cyclist_injured   latitude  \
0                Animals Action                          1        NaN   
1                Animals Action                          1        NaN   
2                Animals Action                          1  40.798290   
3                Animals Action                          1  40.779940   
4                Animals Action                          1  40.640568   

   longitude  
0        NaN  
1        NaN  
2 -73.951840  
3 -73.971030  
4 -73.969167

 How many motor vehicle accidents, injuries, and deaths are there per year in New York City?


In [12]:
query6 = """SELECT
  EXTRACT(YEAR FROM timestamp) AS Year,
  COUNT(*) AS Accidents,
  SUM( number_of_persons_injured ) AS Injuries,
  SUM( number_of_persons_killed) AS Deaths
FROM
  `bigquery-public-data.new_york.nypd_mv_collisions`
GROUP BY
  Year
ORDER BY
  Year ASC;
        """
response6 = nyc.query_to_pandas_safe(query6, max_gb_scanned=10)
response6.head(15)

Year  Accidents  Injuries  Deaths
0  2012     100537     27451     137
1  2013     203720     55121     297
2  2014     206019     51216     262
3  2015     217657     51354     243
4  2016     227808     58888     237
5  2017     229162     60043     247
6  2018      12037      2826       7

![](https://cloud.google.com/blog/big-data/2017/01/images/148467900588042/nyc-dataset-4.png)
https://cloud.google.com/blog/big-data/2017/01/images/148467900588042/nyc-dataset-4.png

What causes these types of motor vehicle accidents?

In [13]:
query7 = """SELECT
  contributing_factor_vehicle_1,
  COUNT(*) AS accidents,
  SUM(number_of_persons_injured) AS injuries,
  SUM(number_of_persons_killed) AS deaths,
  ROUND(100*(SUM( number_of_persons_killed) / COUNT(*)), 3) as pct_causing_death
FROM
  `bigquery-public-data.new_york.nypd_mv_collisions`
GROUP BY
  contributing_factor_vehicle_1
ORDER BY
  accidents DESC,
  pct_causing_death DESC;
        """
response7 = nyc.query_to_pandas_safe(query7, max_gb_scanned=10)
response7.head(15)

contributing_factor_vehicle_1  accidents  injuries  deaths  \
0                      Unspecified     575796    137512     637   
1   Driver Inattention/Distraction     176170     45894     159   
2    Failure to Yield Right-of-Way      57298     23475     105   
3                  Fatigued/Drowsy      48495     12517       3   
4                 Backing Unsafely      37916      3834      19   
5                  Other Vehicular      36038      6922      18   
6            Following Too Closely      31636      9849      17   
7               Turning Improperly      29288      4155       6   
8               Lost Consciousness      20261      2408      10   
9      Traffic Control Disregarded      16190      9680     150   
10         Prescription Medication      15904      3139       7   
11             Driver Inexperience      15637      3659      19   
12  Passing or Lane Usage Improper      14820      1905       8   
13            Unsafe Lane Changing      12472      1783       1   
14               Pavement Slippery      12188      3373       6   

    pct_causing_death  
0               0.111  
1               0.090  
2               0.183  
3               0.006  
4               0.050  
5               0.050  
6               0.054  
7               0.020  
8               0.049  
9               0.926  
10              0.044  
11              0.122  
12              0.054  
13              0.008  
14              0.049

![](https://cloud.google.com/blog/big-data/2017/01/images/148467900588042/nyc-dataset-5.png)
https://cloud.google.com/blog/big-data/2017/01/images/148467900588042/nyc-dataset-5.png

Where do these motor vehicle accidents occur?

In [14]:
query8 = """SELECT
  borough,
  COUNT(*) AS accidents,
  SUM(number_of_persons_killed) AS deaths,
  ROUND(100*SUM(number_of_persons_killed)/COUNT(*), 3) as pct_deaths
FROM
  `bigquery-public-data.new_york.nypd_mv_collisions`
WHERE
  borough != ""
GROUP BY
  borough
ORDER BY
  accidents DESC;
        """
response8 = nyc.query_to_pandas_safe(query8, max_gb_scanned=10)
response8.head(15)

borough  accidents  deaths  pct_deaths
0       BROOKLYN     263715     307       0.116
1         QUEENS     224092     272       0.121
2      MANHATTAN     215819     181       0.084
3          BRONX     114287     130       0.114
4  STATEN ISLAND      39435      54       0.137

![](https://cloud.google.com/blog/big-data/2017/01/images/148467900588042/nyc-dataset-6.png)
https://cloud.google.com/blog/big-data/2017/01/images/148467900588042/nyc-dataset-6.png

What is the ratio of motorist deaths to cyclist deaths?

In [15]:
query9 = """SELECT
  SUM(number_of_motorist_killed) AS motorist_deaths,
  SUM(number_of_cyclist_killed) AS cyclist_deaths,
  SUM(number_of_motorist_killed) / SUM(number_of_cyclist_killed) AS ratio
FROM
  `bigquery-public-data.new_york.nypd_mv_collisions`;
        """
response9 = nyc.query_to_pandas_safe(query9, max_gb_scanned=10)
response9.head(15)

motorist_deaths  cyclist_deaths  ratio
0              544             100   5.44

What’s the Citi Bike record for the Longest Distance in the Shortest Time (on a route with at least 100 rides)?


In [16]:
query10 = """SELECT
  start_station_name,
  end_station_name,
  COUNT(*) trips,
  ROUND(AVG(ACOS( SIN(start_station_latitude*ACOS(-1)/180)*SIN(end_station_latitude*ACOS(-1)/180) + COS(start_station_latitude*ACOS(-1)/180)*COS(end_station_latitude*ACOS(-1)/180) * COS(end_station_longitude*ACOS(-1)/180-start_station_longitude*ACOS(-1)/180) ) * 6371000)) AS distance,
  MIN(tripduration) AS fastest_trip,
  MAX(tripduration) AS slowest_trip,
  ROUND(AVG(tripduration), 3) AS average_trip
FROM
  `bigquery-public-data.new_york.citibike_trips`
WHERE
  start_station_name != end_station_name
  AND start_station_latitude != 0
  AND end_station_latitude != 0
GROUP BY
  start_station_name,
  end_station_name
HAVING
  trips > 100
ORDER BY
  distance DESC
LIMIT
  10;
        """
response10 = nyc.query_to_pandas_safe(query10, max_gb_scanned=10)
response10.head(15)

start_station_name            end_station_name  trips  distance  \
0          Broadway & Battery Pl          W 84 St & Broadway    102    9640.0   
1          Broadway & Battery Pl      Riverside Dr & W 82 St    133    9578.0   
2         Riverside Dr & W 82 St       Broadway & Battery Pl    123    9578.0   
3          Little West St & 1 Pl      Riverside Dr & W 82 St    231    9545.0   
4         Riverside Dr & W 82 St       Little West St & 1 Pl    208    9545.0   
5  Cumberland St & Lafayette Ave            11 Ave & W 59 St    181    9460.0   
6     South End Ave & Liberty St      Riverside Dr & W 82 St    162    8904.0   
7         Riverside Dr & W 82 St  South End Ave & Liberty St    120    8904.0   
8             Broadway & W 58 St     Fulton St & Rockwell Pl    103    8785.0   
9          Broadway & Battery Pl      Riverside Dr & W 72 St    116    8768.0   

   fastest_trip  slowest_trip  average_trip  
0          1775          3788      2361.118  
1          1817          6480      2539.857  
2          1602          4752      2242.000  
3          1607         67970      2995.394  
4          1653          5962      2534.986  
5           604          3376      2298.740  
6          1486          5380      2484.358  
7          1409         25468      2783.950  
8           447          5236      1017.117  
9          1571         13839      2302.629

Are bikes or taxis faster?  Comparing the average duration of 5 of the most popular Citi Bike routes, to taxi journeys beginning and ending within an approximately 50-meter radius of the corresponding Citi Bike stations.



In [17]:
query11 = """SELECT
  start_station_name, end_station_name, avg_bike_duration, avg_taxi_duration, avg_taxi_fare
FROM (
  SELECT
    start_station_name,
    end_station_name,
    ROUND(start_station_latitude, 3) AS ss_lat,
    ROUND(start_station_longitude, 3) AS ss_long,
    ROUND(end_station_latitude, 3) AS es_lat,
    ROUND(end_station_longitude, 3) AS es_long,
    AVG(tripduration) AS avg_bike_duration,
    COUNT(*) AS bike_trips
  FROM
    `bigquery-public-data.new_york.citibike_trips`
  WHERE 
    start_station_name != end_station_name
  GROUP BY
    start_station_name,
    end_station_name,
    ss_lat,
    ss_long,
    es_lat,
    es_long
  ORDER BY
    bike_trips DESC
  LIMIT
    100 )a
JOIN (
  SELECT
    ROUND(pickup_latitude, 3) AS pu_lat,
    ROUND(pickup_longitude, 3) AS pu_long,
    ROUND(dropoff_latitude, 3) AS do_lat,
    ROUND(dropoff_longitude, 3) AS do_long,
    AVG(UNIX_SECONDS(dropoff_datetime)-UNIX_SECONDS(pickup_datetime)) AS avg_taxi_duration,
    AVG(fare_amount) as avg_taxi_fare,
    COUNT(*) AS taxi_trips
  FROM
    `bigquery-public-data.new_york.tlc_yellow_trips_2016`
  GROUP BY
    pu_lat,
    pu_long,
    do_lat,
    do_long)b
ON
  a.ss_lat=b.pu_lat AND
  a.es_lat=b.do_lat AND
  a.ss_long=b.pu_long AND
  a.es_long=b.do_long
ORDER BY
  bike_trips DESC
LIMIT 20;
        """
response11 = nyc.query_to_pandas_safe(query11, max_gb_scanned=10)
response11.head(20)

start_station_name             end_station_name  \
0                    12 Ave & W 40 St        West St & Chambers St   
1                     W 21 St & 6 Ave              9 Ave & W 22 St   
2                     W 21 St & 6 Ave             W 22 St & 10 Ave   
3                      West Thames St     Vesey Pl & River Terrace   
4                   E 7 St & Avenue A        Lafayette St & E 8 St   
5               West St & Chambers St             12 Ave & W 40 St   
6   Grand Army Plaza & Central Park S           Broadway & W 60 St   
7            E 43 St & Vanderbilt Ave              W 41 St & 8 Ave   
8   Grand Army Plaza & Central Park S       Central Park S & 6 Ave   
9                       Old Fulton St      Centre St & Chambers St   
10                   W 22 St & 10 Ave              W 22 St & 8 Ave   
11            Centre St & Chambers St  Cadman Plaza E & Tillary St   
12           Vesey Pl & River Terrace               West Thames St   
13              Pershing Square North              W 33 St & 7 Ave   
14           Vesey Pl & River Terrace    Greenwich St & N Moore St   
15          Greenwich St & N Moore St     Greenwich St & Warren St   
16                    W 17 St & 8 Ave             W 20 St & 11 Ave   
17              Pershing Square North              W 41 St & 8 Ave   
18                   W 22 St & 10 Ave              W 21 St & 6 Ave   
19             Central Park S & 6 Ave           Broadway & W 60 St   

    avg_bike_duration  avg_taxi_duration  avg_taxi_fare  
0         1477.337038        1634.000000      17.250000  
1          326.776957         228.145455       5.040909  
2          422.016836         359.525000       5.837500  
3          460.212957         441.400000       6.500000  
4          341.591626         403.909091       6.181818  
5         1477.415645         924.000000      13.500000  
6         1593.350834         432.989130       6.336957  
7          423.401979        1102.701299       7.266234  
8         3158.462496         417.416667       6.041667  
9         1840.418319         528.500000       8.750000  
10         228.138954         145.861111       5.083333  
11        1770.654200         900.500000      11.750000  
12         479.517693         295.500000       5.500000  
13         518.718877         734.593138       7.705757  
14         373.048328         350.000000       6.071429  
15         267.792917         132.689655       3.603448  
16         348.703906         336.173333       5.540000  
17         478.190632         503.465753       6.893836  
18         435.734608         267.111111       5.013889  
19        2047.676189         266.317460       5.059524

![](https://cloud.google.com/blog/big-data/2017/01/images/148467900588042/nyc-dataset-7.png)
https://cloud.google.com/blog/big-data/2017/01/images/148467900588042/nyc-dataset-7.png

What is the number of 311 requests related to ice cream trucks?


In [18]:
query12 = """SELECT
  COUNT(*) AS requests
FROM
  `bigquery-public-data.new_york.311_service_requests`
WHERE
  LOWER(descriptor) LIKE '%ice cream truck%';
        """
response12 = nyc.query_to_pandas_safe(query12, max_gb_scanned=10)
response12.head(20)

requests
0     12659

Which boroughs have the most 311 requests related to parties?


In [19]:
query13 = """SELECT
  borough,
  COUNT(descriptor) AS parties
FROM
  `bigquery-public-data.new_york.311_service_requests`
WHERE
  LOWER(descriptor) LIKE '%party%'
GROUP BY
  borough
ORDER BY
  parties DESC;
        """
response13 = nyc.query_to_pandas_safe(query13, max_gb_scanned=10)
response13.head(5)

borough  parties
0       BROOKLYN   403971
1      MANHATTAN   373242
2          BRONX   275000
3         QUEENS   239078
4  STATEN ISLAND    34048

What days get the most 311 requests related to parties?


In [20]:
query14 = """SELECT
  extract(DAYOFWEEK
  FROM
    created_date) AS party_day,
  borough,
  COUNT(*) AS num_parties
FROM
  `bigquery-public-data.new_york.311_service_requests`
WHERE
  LOWER(descriptor) LIKE '%party%'
GROUP BY
  party_day,
  borough
ORDER BY
  num_parties DESC;
        """
response14 = nyc.query_to_pandas_safe(query14, max_gb_scanned=10)
response14.head(30)

party_day        borough  num_parties
0           1       BROOKLYN       111114
1           7       BROOKLYN       110643
2           7      MANHATTAN        98680
3           1      MANHATTAN        88933
4           1         QUEENS        79136
5           1          BRONX        76331
6           7          BRONX        71942
7           7         QUEENS        68763
8           6      MANHATTAN        55288
9           6       BROOKLYN        50587
10          2       BROOKLYN        39650
11          6          BRONX        37216
12          5      MANHATTAN        36961
13          2      MANHATTAN        33104
14          5       BROOKLYN        32958
15          4      MANHATTAN        31790
16          4       BROOKLYN        30041
17          3       BROOKLYN        28978
18          3      MANHATTAN        28486
19          6         QUEENS        26835
20          2          BRONX        26100
21          5          BRONX        22779
22          2         QUEENS        20985
23          4          BRONX        20655
24          3          BRONX        19977
25          5         QUEENS        16118
26          4         QUEENS        13882
27          3         QUEENS        13359
28          7  STATEN ISLAND        10376
29          1  STATEN ISLAND         8277

What is the most common factor in a motor vehicle collision in NYC?


In [21]:
query15 = """SELECT
  contributing_factor_vehicle_1 AS collision_factor,
  COUNT(*) num_collisions
FROM
  `bigquery-public-data.new_york.nypd_mv_collisions`
WHERE
  contributing_factor_vehicle_1 != "Unspecified"
  AND contributing_factor_vehicle_1 != ""
GROUP BY
  1
ORDER BY
  num_collisions DESC;
        """
response15 = nyc.query_to_pandas_safe(query15, max_gb_scanned=10)
response15.head(15)

collision_factor  num_collisions
0   Driver Inattention/Distraction          176170
1    Failure to Yield Right-of-Way           57298
2                  Fatigued/Drowsy           48495
3                 Backing Unsafely           37916
4                  Other Vehicular           36038
5            Following Too Closely           31636
6               Turning Improperly           29288
7               Lost Consciousness           20261
8      Traffic Control Disregarded           16190
9          Prescription Medication           15904
10             Driver Inexperience           15637
11  Passing or Lane Usage Improper           14820
12            Unsafe Lane Changing           12472
13               Pavement Slippery           12188
14         Outside Car Distraction           10978

What are the most dangerous streets for motor vehicle collisions in NYC?


In [22]:
query15 = """SELECT
  on_street_name,
  SUM(number_of_persons_killed) AS deaths
FROM
  `bigquery-public-data.new_york.nypd_mv_collisions`
WHERE
  on_street_name <> ''
GROUP BY
  on_street_name
ORDER BY
  deaths DESC
LIMIT
  30;
        """
response15 = nyc.query_to_pandas_safe(query15, max_gb_scanned=10)
response15.head(30)

on_street_name  deaths
0               BROADWAY      23
1        ATLANTIC AVENUE      22
2       LINDEN BOULEVARD      16
3        HYLAN BOULEVARD      16
4     BRUCKNER BOULEVARD      15
5               3 AVENUE      15
6     NORTHERN BOULEVARD      15
7     ROCKAWAY BOULEVARD      14
8        FLATBUSH AVENUE      13
9        GRAND CONCOURSE      12
10      QUEENS BOULEVARD      12
11        JAMAICA AVENUE      10
12           BOSTON ROAD       9
13   WOODHAVEN BOULEVARD       9
14           WEST STREET       9
15        WEBSTER AVENUE       9
16   EAST TREMONT AVENUE       8
17       EASTERN PARKWAY       8
18        BEDFORD AVENUE       8
19         OCEAN PARKWAY       8
20         JEROME AVENUE       7
21    WESTCHESTER AVENUE       7
22   METROPOLITAN AVENUE       7
23  NORTH CONDUIT AVENUE       7
24          CANAL STREET       7
25      AMSTERDAM AVENUE       7
26     ASTORIA BOULEVARD       6
27       NOSTRAND AVENUE       6
28           MAIN STREET       6
29     VICTORY BOULEVARD       6

What are the most popular Citi Bike stations?


In [23]:
query16 = """SELECT
  start_station_name,
  start_station_latitude,
  start_station_longitude,
  COUNT(*) AS num_trips
FROM
  `bigquery-public-data.new_york.citibike_trips`
GROUP BY
  1,
  2,
  3
ORDER BY
  num_trips DESC
LIMIT 10;
        """
response16 = nyc.query_to_pandas_safe(query16, max_gb_scanned=10)
response16.head(30)

start_station_name  start_station_latitude  start_station_longitude  \
0     E 17 St & Broadway               40.737050               -73.990093   
1  Lafayette St & E 8 St               40.730287               -73.990765   
2        W 21 St & 6 Ave               40.741740               -73.994156   
3  West St & Chambers St               40.717548               -74.013221   
4  Pershing Square North               40.751873               -73.977706   
5     Broadway & E 14 St               40.734546               -73.990741   
6     Broadway & E 22 St               40.740343               -73.989551   
7        8 Ave & W 33 St               40.751551               -73.993934   
8       W 20 St & 11 Ave               40.746745               -74.007756   
9     Broadway & W 24 St               40.742354               -73.989151   

   num_trips  
0     291615  
1     277060  
2     275348  
3     260911  
4     246181  
5     244420  
6     237394  
7     223970  
8     211840  
9     211713

What are the most popular routes by subscriber type?


In [24]:
query17 = """SELECT
  usertype,
  CONCAT(start_station_name, " to ", end_station_name) as route,
  COUNT(*) as num_trips,
  ROUND(AVG(cast(tripduration as int64) / 60),2) as duration
FROM
  `bigquery-public-data.new_york.citibike_trips`
GROUP BY
  start_station_name, end_station_name, usertype
ORDER BY
  num_trips DESC
LIMIT 10;
        """
response17 = nyc.query_to_pandas_safe(query17, max_gb_scanned=10)
response17.head(30)

usertype                                              route  num_trips  \
0    Customer   Central Park S & 6 Ave to Central Park S & 6 Ave      40009   
1    Customer  Grand Army Plaza & Central Park S to Grand Arm...      15234   
2    Customer  Centre St & Chambers St to Centre St & Chamber...      12466   
3  Subscriber                 W 21 St & 6 Ave to 9 Ave & W 22 St      11594   
4    Customer           Broadway & W 60 St to Broadway & W 60 St      10816   
5  Subscriber                W 21 St & 6 Ave to W 22 St & 10 Ave      10451   
6  Subscriber         E 7 St & Avenue A to Lafayette St & E 8 St       9639   
7  Subscriber         West Thames St to Vesey Pl & River Terrace       9218   
8  Subscriber                 W 17 St & 8 Ave to 8 Ave & W 31 St       8760   
9  Subscriber        E 43 St & Vanderbilt Ave to W 41 St & 8 Ave       8743   

   duration  
0     50.89  
1     52.99  
2     35.65  
3      5.35  
4     52.36  
5      6.95  
6      5.59  
7      6.67  
8      6.43  
9      7.04

What are the most popular citibike routes for women?

In [25]:
query18 = """SELECT
  CONCAT(start_station_name, " to ", end_station_name) AS route,
  COUNT(*) AS num_trips
FROM
  `bigquery-public-data.new_york.citibike_trips`
WHERE
  gender = "female"
  AND CAST(starttime AS string) LIKE '2016%'
GROUP BY
  start_station_name,
  end_station_name
ORDER BY
  num_trips DESC
LIMIT
  5;
        """
response18 = nyc.query_to_pandas_safe(query18, max_gb_scanned=10)
response18.head(30)

route  num_trips
0                W 21 St & 6 Ave to W 22 St & 10 Ave        942
1  N 6 St & Bedford Ave to Wythe Ave & Metropolit...        905
2                 W 21 St & 6 Ave to 9 Ave & W 22 St        809
3  Wythe Ave & Metropolitan Ave to N 6 St & Bedfo...        769
4                W 22 St & 10 Ave to W 21 St & 6 Ave        765

What are the most popular citibike routes for men?

In [26]:
query19 = """SELECT
  CONCAT(start_station_name, " to ", end_station_name) AS route,
  COUNT(*) AS num_trips
FROM
  `bigquery-public-data.new_york.citibike_trips`
WHERE
  gender = "male"
  AND CAST(starttime AS string) LIKE '2016%'
GROUP BY
  start_station_name,
  end_station_name
ORDER BY
  num_trips DESC
LIMIT
  5;
        """
response19 = nyc.query_to_pandas_safe(query19, max_gb_scanned=10)
response19.head(30)

route  num_trips
0       Pershing Square North to W 33 St & 7 Ave       3355
1       Pershing Square North to W 41 St & 8 Ave       3327
2  Pershing Square North to E 24 St & Park Ave S       2923
3    Pershing Square North to Broadway & W 32 St       2664
4             W 21 St & 6 Ave to 9 Ave & W 22 St       2565

How many trips did Yellow taxis take each month in 2015?


In [27]:
query20 = """SELECT
  TIMESTAMP_TRUNC(pickup_datetime,
    MONTH) month,
  COUNT(*) trips
FROM
  `bigquery-public-data.new_york.tlc_yellow_trips_2015`
GROUP BY
  1
ORDER BY
  1;
        """
response20 = nyc.query_to_pandas_safe(query20, max_gb_scanned=10)
response20.head(30)

month     trips
0  2015-01-01 00:00:00+00:00  12748986
1  2015-02-01 00:00:00+00:00  12450521
2  2015-03-01 00:00:00+00:00  13351609
3  2015-04-01 00:00:00+00:00  13071789
4  2015-05-01 00:00:00+00:00  13158262
5  2015-06-01 00:00:00+00:00  12324935
6  2015-07-01 00:00:00+00:00  11562783
7  2015-08-01 00:00:00+00:00  11130304
8  2015-09-01 00:00:00+00:00  11225063
9  2015-10-01 00:00:00+00:00  12315488
10 2015-11-01 00:00:00+00:00  11312676
11 2015-12-01 00:00:00+00:00  11460573

What was the average speed of Yellow taxi trips in 2015?


In [28]:
query21 = """SELECT
  EXTRACT(HOUR
  FROM
    pickup_datetime) hour,
  ROUND(AVG(trip_distance / TIMESTAMP_DIFF(dropoff_datetime,
        pickup_datetime,
        SECOND))*3600, 1) speed
FROM
  `bigquery-public-data.new_york.tlc_yellow_trips_2015`
WHERE
  trip_distance > 0
  AND fare_amount/trip_distance BETWEEN 2
  AND 10
  AND dropoff_datetime > pickup_datetime
GROUP BY
  1
ORDER BY
  1;
        """
response21 = nyc.query_to_pandas_safe(query21, max_gb_scanned=10)
response21.head(30)

hour  speed
0      0   15.8
1      1   16.3
2      2   16.8
3      3   17.5
4      4   20.0
5      5   21.6
6      6   17.6
7      7   13.7
8      8   11.6
9      9   11.4
10    10   11.5
11    11   11.3
12    12   11.2
13    13   11.3
14    14   11.2
15    15   11.0
16    16   11.5
17    17   11.2
18    18   11.1
19    19   11.8
20    20   12.9
21    21   13.6
22    22   14.1
23    23   14.9

What was the average speed of Yellow taxi trips in 2015 by day of week?


In [29]:
query22 = """SELECT
  EXTRACT(DAYOFWEEK
  FROM
    pickup_datetime) DAYOFWEEK,
  ROUND(AVG(trip_distance / TIMESTAMP_DIFF(dropoff_datetime,
        pickup_datetime,
        SECOND))*3600, 1) speed
FROM
  `bigquery-public-data.new_york.tlc_yellow_trips_2015`
WHERE
  trip_distance > 0
  AND fare_amount/trip_distance BETWEEN 2
  AND 10
  AND dropoff_datetime > pickup_datetime
GROUP BY
  1
ORDER BY
  1;
        """
response22 = nyc.query_to_pandas_safe(query22, max_gb_scanned=10)
response22.head(30)

DAYOFWEEK  speed
0          1   14.7
1          2   13.4
2          3   12.3
3          4   12.2
4          5   12.2
5          6   12.3
6          7   13.1